tmpl/wiki/error-handling.md

```diff
@@ -0,0 +1,384 @@
+
+Thirdly, it can evaluate to a returned (successful) thread in all cases, containing a variant indicating success or failure:
+
+```
+let load x = if ...
+  then return (`Error (Permission_denied "Cannot read ..."))
+  else return (`Ok ...)
+```
+
+The first two are very similar, as if the exception raised in the first example reaches Lwt it will be turned into a failed thread.
+The difference is that, depending on exactly how the threads are scheduled, a raised exception might make it back to the caller as an exception or as a failed thread. A buggy caller might fail to handle one or the other case (though using a `try_lwt` block will handle both correctly).
+Therefore, using the second form instead of the first is preferred.
+
+The Mirage interfaces generally prefer the third option.
+Most of the module types in `V1` include an `error` type. Some leave it abstract, while others define it in various ways:
+
+- `RANDOM` and `FLOW` leave it abstract.
+- `ENTROPY`, `CONSOLE`, `BLOCK`, `NETWORK`, `ETHIF`, `IP`, `UDP`, `TCP` and `STACKV4` define it as a polymorphic variant.
+- `KV_RO` defines it as a regular variant.
+- `FS` defines it, but in terms of the abstract `block_device_error` type.
+
+In theory, explicitly listing all possible errors and forcing callers to consider them should result in more reliable code.
+In practice, it results in callers ignoring all errors.
+
+The first problem is that there is no easy way to display an error to the user.
+If you look inside any `main.ml` file you will see code like this:
+
+```
+let t3 () =
+  console1 () >>= function
+  | `Error e -> fail (Failure "console1")
+  | `Ok console1 ->
+  block1 () >>= function
+  | `Error e -> fail (Failure "block1")
+  | `Ok block1 ->
+  return (`Ok (console1, block1))
+```
+
+Here, if the console or block devices return an error then we throw it away and report a generic error which just says which device failed, but not why.
+This approach to error handling can be seen throughout the Mirage code (grepping for `Error _` will turn up quite a few).
+Here's an example from `mirage-skeleton/block/unikernel.ml`, teaching new users how to redefine the bind operator to discard error details:
+
+```
+let ( >>= ) x f = x >>= function
+  | `Error _ -> fail (Failure "error")
+  | `Ok x -> f x
+```
+
+Another problem is that we force callers explicitly to handle errors that they would never want to handle.
+For example, in `BLOCK` we have:
+
+```
+module type BLOCK = sig
+  type error = [
+    | `Unknown of string (** an undiagnosed error *)
+    | `Unimplemented     (** operation not yet implemented in the code *)
+    | `Is_read_only      (** you cannot write to a read/only instance *)
+    | `Disconnected      (** the device has been previously disconnected *)
+  ]
+
+  val read:
+    t -> int64 -> page_aligned_buffer list ->
+    [ `Error of error | `Ok of unit ] io
+
+  ...
+```
+
+The implication is that callers of `read` should write code to handle each case.
+But how?
```

Comment:
Not really. You could have another binding function (&>>=) that would let you handle only the `Ok case and propagate the error up without having you to bother about it except at the top level.